WrapImage/nifti_wrapper.py

```diff
@@ -75,52 +24,162 @@ def loop_over_first_n_minus_1_dimensions(arr):
         flat_view = arr[idx].flatten()
         yield idx, flat_view
 
+def generate_data(data, bvals, b0_indices, groups, total_iteration):
+    num_directions = groups.shape[1]
+    data = data.reshape(total_iteration, -1)
+    for idx in range(total_iteration):
+        for dir in range(num_directions):
+            # print('yielding')
+            yield (data[idx, groups[:, dir]].flatten(), bvals[:, groups[:, dir]].ravel(), b0_indices[:, groups[:, dir]].ravel())
+
+def osipi_fit(fitfunc, data_bvals):
+    data, bvals, b0_indices = data_bvals
```

Comment:
Does it make more sense to use dictionaries for these?

Reply:
Yes, I think Ivan has a PR for that so maybe there's a conflict between these. I guess either his or mine should get merged first and then the other will need to update the other.